compiler_gym/envs/cgra/DFG.py

```diff
@@ -114,9 +114,30 @@ def bfs(self):
             seen.add(head)
             yield self.nodes[head]
 
-            # Get the following nodes.
-            following_nodes = self.adj[head]
-            to_explore += following_nodes
+            # Add the next batch of nodes that we have
+            # visited all the preds for if there are more
+            # nodes to explore.
+            if len(to_explore) == 0 and len(seen) != len(self.node_names):
+                for node_name in self.node_names:
+                    if node_name in seen:
+                        continue
+                    else:
+                        # Unseen --- have we seen all th preds?
+                        failed = False
+                        for p in preds_lookup[node_name]:
+                            if p.name not in seen:
+                                failed = True
+                        if not failed:
+                            to_explore.append(node_name)
+                if len(to_explore) == 0: # We added nothing despite trying
+                    # to.
+
+                    # TODO(jcw) -- Fix this, as support for cyclical DFGs
+                    # is important to be able to support loops with 
+                    # cross-loop dependencies.
+                    print("Cyclical DFG --- Impossible to do a true BFS")
+                    print("DFG is ", str(self))
+                    assert False
```

Comment:
Nitpick: don't assert False, raise an exception. Or, if that's not possible, `sys.exit(1)`.